### PR TITLE
[next] Separate contexts by domain

### LINKED ISSUE #998: [next] Domain-specific contexts
Currently, Threlte uses two contexts, one of which is part of the public API.

In an effort to support advanced patterns like `<HUD>` or `useViewport` we want to separate these two contexts to a multitude of contexts which can be created at any point in the component tree. We're therefore opting for domain-specific contexts:

- Scene Context
	- scene
- Camera Context
	- camera
- Renderer Context
	- colorSpace
	- toneMapping
	- dpr
	- shadows
	- colorManagementEnabled
	- useLegacyLights
	- renderer
- Scheduler Context
	- scheduler
	- renderMode
	- autoRender
	- frameInvalidated
	- advance
	- autoInvalidations
	- resetFrameInvalidation
	- mainStage
	- renderStage
	- autoRenderTask
	- shouldRender
- Disposal Context
	- dispose
	- collectDisposableObjects
	- addDisposableObjects
	- removeDisposableObjects
	- disposableObjects
	- shouldDispose
- Cache Context
	- remember
	- clear
- (Plugins Context)
	- plugins

Every context has at least two methods:

- `createXContext` to create the context
  - `(args) => ContextType`
- `useX` to use the context
  - `() => ContextType`

On top of that, there will be convenience functions that create *all* contexts at once:

- `createThrelteContext`

This PR intentionally leaves out the size management that Threlte currently does as it will not be used in Threlte 8.

### PR DESCRIPTION
**Pull request by grischaerbe**
Fixes #998 

This PR also addresses a flaw in the design of the `<Canvas>` component: Since the `<Canvas>` markup only contains a `<canvas>` element, there's no way to have authority over added HTML elements, since they cannot be children to a `<canvas>` element. This PR introduces a new layout:

```html
<div>
  <canvas></canvas>
</div>
```

which gives us authority over elements being added as children to the `<div>` element. This paves the way for #809, #849, #949

### CODE REVIEW

File: packages/core/src/lib/context/fragments/renderer.svelte.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  if ('setAnimationLoop' in context.renderer) {
+    const setAnimationLoop = context.renderer.setAnimationLoop as (
+      callback: (time: number) => void
+    ) => void
+    setAnimationLoop((time) => {
+      dispose()
+      scheduler.run(time)
+      resetFrameInvalidation()
+    })
+  }
```
Comment (michealparks) [inline packages/core/src/lib/context/fragments/renderer.svelte.ts:170]: When is this not true?

File: packages/core/src/lib/context/fragments/renderer.svelte.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  onDestroy(() => {
+    if ('dispose' in context.renderer) {
+      const dispose = context.renderer.dispose as () => void
+      dispose()
+    }
+  })
+
+  $effect(() => {
+    context.colorManagementEnabled.set(options.colorManagementEnabled ?? true)
+  })
```
Comment (michealparks) [inline packages/core/src/lib/context/fragments/renderer.svelte.ts:181]:
```suggestion
```
duplicate ✂️

File: packages/core/src/lib/context/fragments/renderer.svelte.ts (hunk 1 above)
Comment (grischaerbe) [reply packages/core/src/lib/context/fragments/renderer.svelte.ts:170]: `context.renderer` is now extending the generic `THREE.Renderer` which currently is this:

```ts
export interface Renderer {
    domElement: HTMLCanvasElement;

    render(scene: Object3D, camera: Camera): void;
    setSize(width: number, height: number, updateStyle?: boolean): void;
}
```

Comment (michealparks) [reply packages/core/src/lib/context/fragments/renderer.svelte.ts:170]: Ah!

**Review by michealparks — APPROVED** [review]
😍

**Pull request merged**